### PR TITLE
[location] expo-location audit and refactor

### LINKED ISSUE #4015: getCurrentPositionAsync does not work concurrently
### Environment

Expo 32.0

### Steps to Reproduce

Insert the following code in an Expo project and run it.

```javascript
import { Permissions, Location } from 'expo';

Permissions.askAsync(Permissions.LOCATION).then(({ status }) => {
  if (status === "granted") {
    console.log("Asking for location twice");

    Location.getCurrentPositionAsync({
      accuracy: Location.Accuracy.Highest
    }).then((l) => console.log("First", l));

    Location.getCurrentPositionAsync({
      accuracy: Location.Accuracy.Highest
    }).then((l) => console.log("Second", l));
  }
});
```

### Expected Behavior

Both promises from the two calls to `getCurrentPositionAsync` resolve and both `First` and `Second` is logged to the console.

### Actual Behavior

Only `Second` is ever logged to the console. The promise from the first invocation to `getCurrentPositionAsync` never resolves.

Note that it is always `Second` that gets logged and never `First`. It appears to me that when `getCurrentPositionAsync` is called while there already is an ongoing location request that hasn't resolved yet the promise from the first call is forgotten and never resolved.

### Reproducible Demo

Here is a Snap reproducing the issue. I've simply pasted the code above into the default project Snap creates: https://snack.expo.io/SJGdW6E5E

### LINKED ISSUE #4304: Android Location.getCurrentPositionAsync as per sdk32 docs not working, but as per sdk31 docs is working
## 🐛 Bug Report
Android Location.getCurrentPositionAsync as per sdk32 docs not working, but as per sdk31 docs is working
### Environment
SDK 32, Android

  Expo CLI 2.18.3 environment info:
    System:
      OS: macOS 10.14.4
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.15.3 - /usr/local/bin/node
      npm: 6.4.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.3 AI-182.5107.16.33.5314842
      Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^32.0.6 => 32.0.6
      react: ^16.5.0 => 16.5.0
      react-native: https://github.com/expo/react-native/archive/sdk-32.0.1.tar.gz => 0.57.1
      react-navigation: ^3.4.1 => 3.4.1
    npmGlobalPackages:
      expo-cli: 2.18.3

### Steps to Reproduce
Working Code:
let position = await Location.getCurrentPositionAsync({enableHighAccuracy:true});
Not Working Code:
let position = await Location.getCurrentPositionAsync({accuracy:Location.Accuracy.Lowest});

### LINKED ISSUE #5249: [ web ] LocationOptions in web are different from native
## 🐛 Bug Report

The documented options for `expo-location` are here: https://docs.expo.io/versions/latest/sdk/location/#arguments-1

however these are not currently accurate for web, as it passes the options directly to the browsers `Geolocation.getCurrentPosition()` (https://developer.mozilla.org/en-US/docs/Web/API/PositionOptions)

This leads to needing to use `maximumAge` on web and `timeInterval` on native.

Ideally, when possible these should match up:

- `timeInterval`  should pass to `maximumAge` on web
- `accuracy` should pass to `enableHighAccuracy` on web

I'm not sure how best to handle the other options as they only exist on web or mobile.

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

`expo-location` wins the challenge for the number of issues being opened 😓 

Fixes #4015
Fixes #4304
Fixes #5249

# How

I've decided to do an audit over it which should result in fixing some of those reported bugs and generally improve its reliability and make API more consistent across platforms.

- `getLastKnownPositionAsync`
  - Now doesn't request for the new location on iOS
  - No longer rejects when the last known location is not available, returns `null` instead
- Removed deprecated `enableHighAccuracy`
- Removed `maximumAge` and `timeout` options from `getCurrentPositionAsync` — they were Android only and honestly the same behavior can be implemented on the JS side for all platforms so I see no reason to maintain this code on the native side
- Some of TypeScript types have changed their names — it might be considered breaking, but this is a change for the better
  - `LocationData` -> `LocationObject`
  - `Accuracy` -> `LocationAccuracy`
  - `ProviderStatus` -> `LocationProviderStatus`
  - `HeadingData` -> `LocationHeadingObject`
  - ... and more like this ...
- Added some new tests
- Improved Web support
  - Added missing methods (for permissions and last known location)
  - Make types consistent with other platforms
  - Enabled web-compatible tests in test-suite
- Better handling of `watchPositionAsync` and `watchHeadingAsync` subscriptions
- `altitudeAccuracy` now available on Android as well (at least Android 8.0)

# Test Plan

- Ran through `test-suite` tests on Android, iOS and Web – all is green ✅ 

# To do later

- I'm wondering whether it's a good chance to rename enums to be all UPPER_CASED as we use this convention right now 🤔

### CODE REVIEW

**Comment by tsapeta** [comment]
I already requested for review, but I will understand if you guys think it's too big to go through 😅 Sorry for that, it was unintentional but that's ok to me to split it up if you would need to.

File: packages/expo-location/src/Location.types.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+/**
+ * Represents subscription object returned by methods watching for new locations or headings.
+ */
+export type LocationSubscription = {
+  remove: () => void;
+};
+
+export type PermissionDetailsLocationIOS = {
+  scope: 'whenInUse' | 'always';
```
Comment (lukmccall) [inline packages/expo-location/src/Location.types.ts:197]:
```suggestion
  scope: 'whenInUse' | 'always' | 'none';
```

**Review by lukmccall — APPROVED** [review]
I left some comments ;) but I think I will read this code one more time :D

**Comment by lukmccall** [comment]
One more thing 😅

Currently, you added `isLocationValid` to every platform. I think we could get the last know position from the native side and validate through js code. In that way, we don't need to add the `isLocationValid` method to every platform. Moreover, in that case, we don't need to pass options to native code either.

**Review by lukmccall — CHANGES_REQUESTED** [review]
One more thing 😅

Currently, you added `isLocationValid` to every platform. I think we could get the last know position from the native side and validate through js code. In that way, we don't need to add the `isLocationValid` method to every platform. Moreover, in that case, we don't need to pass options to native code either.

File: docs/pages/versions/unversioned/sdk/location.md [hunk 2]
```diff
@@ -114,57 +114,59 @@ Asks the user to grant permissions for location. Alias for `Permissions.askAsync
...
 #### Arguments
 
-- **options (_object_)** -- A map of options:
+- **options (_object_)** — A map of options:
 
-  - **accuracy : [Location.Accuracy](#locationaccuracy)** -- Location manager accuracy. Pass one of [Location.Accuracy](#locationaccuracy) enum values. For low accuracy the implementation can avoid geolocation providers that consume a significant amount of power (such as GPS).
-  - **timeInterval (_number_)** -- Minimum time to wait between each update in milliseconds.
-  - **distanceInterval (_number_)** -- Receive updates only when the location has changed by at least this distance in meters.
-  - **mayShowUserSettingsDialog (_boolean_)** -- Specifies whether to ask the user to turn on improved accuracy location mode which uses Wi-Fi, cell networks and GPS sensor. The dialog can be shown only when the location mode is set to **Device only**. Defaults to `true`. (**Android only**)
+  - **accuracy : [LocationAccuracy](#locationaccuracy)** — Location manager accuracy. Pass one of [LocationAccuracy](#locationaccuracy) enum values. For low accuracy the implementation can avoid geolocation providers that consume a significant amount of power (such as GPS).
```
[the remark this answers is not in the document]
Comment (tsapeta) [reply docs/pages/versions/unversioned/sdk/location.md:164]: Yeah, I've used em dash `—` (`-`, `Option`, `Shift`) for this which is for more more natural here, even more than single short dash. I think it's a good idea to change them all, but maybe not in this PR? 🤔

**Comment by tsapeta** [comment]
@lukmccall I think it's a good point and I've been thinking about it, but for me it's always more clear to have the main implementation in one place and don't have to worry about switching between different contexts (code editors, languages and so on). Imho JS side should only be an interface to the native API (except web of course). Anyway, I can change this if you're still not convinced 😄

**Review by lukmccall — APPROVED** [review]

**Pull request merged**